src/main/java/org/reaktivity/nukleus/tcp/internal/writer/stream/Slab.java

```diff
@@ -0,0 +1,170 @@
+
+    Slab(int maximumPartiallyWrittenStreams, int maximumWindowSize)
+    {
+        validateGT0(maximumPartiallyWrittenStreams, "maximumPartiallyWrittenStreams");
+        validateGT0(maximumWindowSize, "maximumWindowSize");
+        int slots = findNextPositivePowerOfTwo(maximumPartiallyWrittenStreams);
+        slotSize = findNextPositivePowerOfTwo(maximumWindowSize);
+        slotSizeBits = Integer.numberOfTrailingZeros(slotSize);
+        int capacityInBytes = slots << slotSizeBits;
+        buffer = ByteBuffer.allocateDirect(capacityInBytes);
+        writeBuffer = ByteBuffer.allocateDirect(slotSize);
+        offsets = new int[slots];
+        Arrays.fill(offsets, UNUSED);
+        remaining = new int[slots];
+    }
+
+    public ByteBuffer get(int slot, DirectBuffer data, int offset, int length)
+    {
+        ByteBuffer result;
+        if (slot == NO_SLOT)
+        {
+            writeBuffer.clear();
+            data.getBytes(offset, writeBuffer, length);
+            writeBuffer.flip();
+            result = writeBuffer;
+        }
+        else
+        {
+            // Append the data to the previous remaining data
+            buffer.position(offsets[slot] + remaining[slot]);
+            buffer.limit((slot << slotSizeBits) + slotSize);
+            data.getBytes(offset, buffer, length);
+            remaining[slot] += length;
+            buffer.position(offsets[slot]);
+            buffer.limit(buffer.position() + remaining[slot]);
+            result = buffer;
+        }
+        return result;
+    }
+
+    public ByteBuffer get(int slot)
+    {
+        ByteBuffer result = null;
+        if (slot != NO_SLOT)
+        {
+            buffer.position(offsets[slot]);
+            buffer.limit(offsets[slot] + remaining[slot]);
+            result = buffer;
+        }
+        return result;
+    }
+
+    public int written(long streamId, int slot, ByteBuffer written)
+    {
+        int nextSlot = slot;
+        if (slot == NO_SLOT)
+        {
+            if (written.hasRemaining())
+            {
+                // store the remaining data into a new slot
+                nextSlot = acquire(streamId);
+                if (nextSlot != OUT_OF_MEMORY)
+                {
+                    remaining[nextSlot] = written.remaining();
+                    buffer.limit(offsets[nextSlot] + slotSize);
+                    buffer.position(offsets[nextSlot]);
+                    buffer.put(written);
+                    buffer.limit(buffer.position());
+                    buffer.position(offsets[nextSlot]);
+                }
+            }
+            // else: nothing to do, continue to use the default writeBuffer
+        }
+        else
+        {
+            if (written.hasRemaining())
+            {
+                // Some data from the existing slot was written, adjust offset and remaining
+                offsets[slot] = written.position();
+                remaining[slot] = written.remaining();
```

Comment:
In principle it can't happen since the slot size is the max window size. Writer Stream.processData with its call to reduceWindow guards against the client violating the window size limit.